File: sudoku.jl
#!/usr/bin/env julia

using IterTools
using ArgParse

# SudokuField() = IntSet([1,2,3,4,5,6,7,8,9])
# SudokuField(n) = (1 <= n <= 9) ? IntSet([n]) : error("invalid sudoku field")
# field_is_fixed(f) = (possibilities(f) == 1)
# remove_options(f, vals) = setdiff(f, vals)
# impose_num!(sudoku, i, j, f::IntSet) = impose_num!(sudoku, i, j, collect(f)[1])
# get_num(f) = collect(f)[1]
# possibilities(f) = length(f)
# get_possibilities(f) = f

SudokuField() = 0x1ff
SudokuField(n) = (1 <= n <= 9) ? (1 << (n-1)) : error("invalid sudoku field ", n)
field_is_fixed(f) = (
    f == 0 ? false : ((f & 1) == 1 ? f == 1 : field_is_fixed((f>>1))))
remove_options(f, vals::UInt16) = f & ~vals

function remove_options(f, vals::Array{T}) where {T<:Integer}
    vals_bits::UInt16 = 0
    for v in vals
        vals_bits |= 1 << (v-1)
    end
    return remove_options(f, vals_bits)
end

function get_num(f)
    for n in 1:9
        if (f & (1<<(n-1))) == f
            return n
        end
    end
    return 0
end

possibilities(f) = (f == 0 ? 0 : possibilities(f>>1) + (f&1))

get_possibilities(f) = [n for n in 1:9 if (f & (1<<(n-1))) != 0]

impose_num!(sudoku, i, j, f::UInt16) = impose_num!(sudoku, i, j, get_num(f))

Sudoku() = fill(SudokuField(), (9,9))

row_of(i,j) = ((i,k) for k in 1:9)
col_of(i,j) = ((k,j) for k in 1:9)
block_of(i,j) = ((k,l) for k in 3*div(i-1,3)+1:3*div(i-1,3)+3
                       for l in 3*div(j-1,3)+1:3*div(j-1,3)+3)

function impose_num!(sudoku, i, j, n::Integer)
    newly_fixed = Set{Tuple{Int,Int}}()

    sudoku[i,j] = SudokuField(n)

    @itr for (k,l) in chain(col_of(i,j), row_of(i,j), block_of(i,j))
        if k == i && j == l
            continue
        end
        was_fixed = field_is_fixed(sudoku[k,l])
        sudoku[k,l] = remove_options(sudoku[k,l], [n])
        if !was_fixed && field_is_fixed(sudoku[k,l])
            push!(newly_fixed, (k,l))
        end
    end

    for (k,l) in newly_fixed
        if !field_is_fixed(sudoku[k,l])
            continue
        end
        impose_num!(sudoku, k, l, sudoku[k,l])
    end
end

function eliminate_in_region!(sudoku, indexpairs)
    imposed = 0

    for (i,j) in indexpairs
        f = sudoku[i,j]

        if field_is_fixed(f)
            continue
        end

        for (k,l) in indexpairs
            if k == i && l == j
                continue
            end
            f = remove_options(f, sudoku[k,l])
        end

        if possibilities(f) == 1
            impose_num!(sudoku, i, j, f)
            imposed += 1
        end
    end

    return imposed
end

function do_elimination!(sudoku)
    imposed = 0
    for i in 1:9
        imposed += eliminate_in_region!(sudoku, collect(row_of(i, i)))
        imposed += eliminate_in_region!(sudoku, collect(col_of(i, i)))
        imposed += eliminate_in_region!(sudoku, collect(block_of(((i-1)%3)*3+1, div(i-1,3)*3+1)))
    end

    if imposed != 0
        do_elimination!(sudoku)
    end
end

function solve!(sudoku, find_all_solutions=true, callback=nothing)
    do_elimination!(sudoku)

    status = check_solution(sudoku)
    if status == SUDOKU_DONE
        if callback != nothing
            callback(sudoku)
        end
        return 1
    elseif status == SUDOKU_ERROR
        return 0
    else
        # We'll have to do some guessing

        # Find the lowest-order point
        lowest_order = 9
        lowest_i = 0
        lowest_j = 0
        for j in 1:9
            for i in 1:9
                f = sudoku[i,j]
                if 1 < possibilities(f) < lowest_order
                    lowest_order = possibilities(f)
                    lowest_i = i
                    lowest_j = j
                    if possibilities(f) == 2
                        break
                    end
                end
            end
        end

        options = sudoku[lowest_i,lowest_j]
        solution = nothing
        solutions_count = 0

        for n in get_possibilities(options)
            trial_sudoku = copy(sudoku)
            impose_num!(trial_sudoku, lowest_i, lowest_j, n)
            solutions_here = solve!(trial_sudoku, find_all_solutions, callback)
            solutions_count += solutions_here
            if solutions_here > 0
                solution = copy(trial_sudoku)

                if !find_all_solutions
                    break
                end
            end
        end

        if solutions_count != 0
            sudoku[:] = solution[:]
        end
        return solutions_count
    end
end

@enum SudokuStatus SUDOKU_ERROR SUDOKU_DONE SUDOKU_IN_PROGRESS

function check_solution(sudoku)
    any_not_fixed = false
    for f in sudoku
        if possibilities(f) == 0
            return SUDOKU_ERROR
        elseif !field_is_fixed(f)
            any_not_fixed = true;
        end
    end
    return any_not_fixed ? SUDOKU_IN_PROGRESS : SUDOKU_DONE
end


function Sudoku(nums::Matrix{T}) where {T<:Integer}
    if size(nums) != (9,9)
        error("sudoku are 9x9")
    end

    s = Sudoku()
    for j in 1:9
        for i in 1:9
            n = nums[i,j]
            if 1 <= n <= 9
                impose_num!(s, i, j, n)
            end
        end
    end
    return s
end

function read_nums(stream::IO)
    m = zeros(Int, (9,9))
    for i in 1:9
        for j in 1:9
            while true
                c = 0
                try
                    c = read(stream, Char)
                catch EOFError
                    return m
                end

                if isdigit(c)
                    m[i,j] = parse(Int, c)
                    break
                elseif isspace(c)
                    # skip spaces
                else
                    break
                end
            end
        end
    end
    return m
end

read_nums(s::String) = read_nums(IOBuffer(s))

function print_sudoku(s, short=false)
    for i in 1:9
        for j in 1:9
            f = s[i,j]
            print(field_is_fixed(f) ? get_num(f) : '.')
            if !short
                print(j == 9 ? '\n' : ' ')
            end
        end
    end
end

function main()
    s = ArgParseSettings()
    @add_arg_table s begin
        "--count-solutions", "-c"
            help = "Count how many solutions there are (default)"
            dest_name = "count-solutions"
            action = :store_true
        "--do-not-count", "-C"
            help = "Do not count how many solutions there are"
            dest_name = "count-solutions"
            action = :store_false
            force_override = true
        "--all", "-a"
            help = "Find all solutions to the puzzle"
            action = :store_true
        "--short-output", "-s"
            help = "Use a shorter output format"
            action = :store_true
        "--timeit"
            help = "Time the solver"
            metavar = "ITERATIONS"
            arg_type = Int
            default = 0
        "sudoku_file"
            help = "input file(s)"
            nargs = '*'

    end

    args = parse_args(s)

    my_print_sudoku(s) = print_sudoku(s, args["short-output"])

    sudoku_files = args["sudoku_file"]
    if length(sudoku_files) == 0
        sudoku_files = ["-"]
    end

    for fn in sudoku_files
        fp = STDIN
        if fn != "-"
           fp = open(fn, "r") 
        end

        while true

            sudoku_nums = read_nums(fp)
            if all(sudoku_nums .== 0)
                exit(0)
            end

            s = Sudoku(sudoku_nums)
            if !args["short-output"]
                println("Sudoku:")
                my_print_sudoku(s)
                print("\n")
            end

            if args["timeit"] > 0
                t1 = time() 
                n = 0
                for i in 1:args["timeit"]
                    s = Sudoku(sudoku_nums)
                    n = solve!(s, args["count-solutions"])
                end
                t2 = time()
                dt_ms = (t2 - t1)*1000
                
                if args["short-output"]
                    my_print_sudoku(s)
                    if args["count-solutions"]
                        @printf " %d" n
                    end
                    @printf " %f\n" dt_ms/args["timeit"]
                else
                    if args["count-solutions"]
                        if n > 1
                            @printf "\nThere are %d solutions.\n" n
                        elseif n == 1
                            print("\nThere is 1 solution.\n")
                        else
                            print("\nThere are no solutions.\n")
                        end
                    end
                    my_print_sudoku(s)
                    @printf "Running time %.2f s (%.2f ms per iteration)\n" dt_ms/1e3 dt_ms/args["timeit"]
                end
            else
                n = 0
                if args["all"]
                    n = solve!(s, true, args["short-output"] ? nothing : my_print_sudoku)
                else
                    n = solve!(s, args["count-solutions"])
                end

                if args["short-output"]
                    if args["count-solutions"]
                        my_print_sudoku(s)
                        @printf " %d\n" n
                    else
                        my_print_sudoku(s)
                        println("")
                    end
                else
                    if args["count-solutions"]
                        if n > 1
                            @printf "There are %d solutions.\n" n
                        elseif n == 1
                            print("There is 1 solution.\n")
                        else
                            print("There are no solutions.\n")
                        end
                    end
                    if !args["all"]
                        my_print_sudoku(s)
                    end
                end
            end
        end
    end
end

main()

File: sudoku.py
#!/usr/bin/env python3


#############################################
# START OF NORVIG CODE
# From https://github.com/norvig/pytudes/blob/master/sudoku.py
#############################################


## Solve Every Sudoku Puzzle

## See http://norvig.com/sudoku.html

## Throughout this program we have:
##   r is a row,    e.g. 'A'
##   c is a column, e.g. '3'
##   s is a square, e.g. 'A3'
##   d is a digit,  e.g. '9'
##   u is a unit,   e.g. ['A1','B1','C1','D1','E1','F1','G1','H1','I1']
##   grid is a grid,e.g. 81 non-blank chars, e.g. starting with '.18...7...
##   values is a dict of possible values, e.g. {'A1':'12349', 'A2':'8', ...}

def cross(A, B):
    "Cross product of elements in A and elements in B."
    return [a+b for a in A for b in B]

digits   = '123456789'
rows     = 'ABCDEFGHI'
cols     = digits
squares  = cross(rows, cols)
unitlist = ([cross(rows, c) for c in cols] +
            [cross(r, cols) for r in rows] +
            [cross(rs, cs) for rs in ('ABC','DEF','GHI') for cs in ('123','456','789')])
units = dict((s, [u for u in unitlist if s in u])
             for s in squares)
peers = dict((s, set(sum(units[s],[]))-set([s]))
             for s in squares)

################ Unit Tests ################

def test():
    "A set of tests that must pass."
    assert len(squares) == 81
    assert len(unitlist) == 27
    assert all(len(units[s]) == 3 for s in squares)
    assert all(len(peers[s]) == 20 for s in squares)
    assert units['C2'] == [['A2', 'B2', 'C2', 'D2', 'E2', 'F2', 'G2', 'H2', 'I2'],
                           ['C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9'],
                           ['A1', 'A2', 'A3', 'B1', 'B2', 'B3', 'C1', 'C2', 'C3']]
    assert peers['C2'] == set(['A2', 'B2', 'D2', 'E2', 'F2', 'G2', 'H2', 'I2',
                               'C1', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9',
                               'A1', 'A3', 'B1', 'B3'])
    print('All tests pass.')

################ Parse a Grid ################

def parse_grid(grid):
    """Convert grid to a dict of possible values, {square: digits}, or
    return False if a contradiction is detected."""
    ## To start, every square can be any digit; then assign values from the grid.
    values = dict((s, digits) for s in squares)
    for s,d in grid_values(grid).items():
        if d in digits and not assign(values, s, d):
            return False ## (Fail if we can't assign d to square s.)
    return values

def grid_values(grid):
    "Convert grid into a dict of {square: char} with '0' or '.' for empties."
    chars = [c for c in grid if c in digits or c in '0.']
    assert len(chars) == 81
    return dict(zip(squares, chars))

################ Constraint Propagation ################

def assign(values, s, d):
    """Eliminate all the other values (except d) from values[s] and propagate.
    Return values, except return False if a contradiction is detected."""
    other_values = values[s].replace(d, '')
    if all(eliminate(values, s, d2) for d2 in other_values):
        return values
    else:
        return False

def eliminate(values, s, d):
    """Eliminate d from values[s]; propagate when values or places <= 2.
    Return values, except return False if a contradiction is detected."""
    if d not in values[s]:
        return values ## Already eliminated
    values[s] = values[s].replace(d,'')
    ## (1) If a square s is reduced to one value d2, then eliminate d2 from the peers.
    if len(values[s]) == 0:
        return False ## Contradiction: removed last value
    elif len(values[s]) == 1:
        if not all(eliminate(values, s2, values[s]) for s2 in peers[s]):
            return False
    ## (2) If a unit u is reduced to only one place for a value d, then put it there.
    for u in units[s]:
        dplaces = [s for s in u if d in values[s]]
        if len(dplaces) == 0:
            return False ## Contradiction: no place for this value
        elif len(dplaces) == 1:
            # d can only be in one place in unit; assign it there
            if not assign(values, dplaces[0], d):
                return False
    return values


################ Search ################

def solve(grid):
    return search(parse_grid(grid))

def search(values):
    "Using depth-first search and propagation, try all possible values."
    if values is False:
        return False ## Failed earlier
    if all(len(values[s]) == 1 for s in squares):
        return values ## Solved!
    ## Chose the unfilled square s with the fewest possibilities
    n,s = min((len(values[s]), s) for s in squares if len(values[s]) > 1)
    return some(search(assign(values.copy(), s, d)) for d in values[s])

################ Utilities ################

def some(seq):
    "Return some element of seq that is true."
    for e in seq:
        if e: return e
    return False

def from_file(filename, sep='\n'):
    "Parse a file into a list of strings, separated by sep."
    return file(filename).read().strip().split(sep)

#############################################
# END OF NORVIG CODE
#############################################


# command-line UI vaguely compatible with the other language versions

import sys
import argparse
import time

def print_grid(grid, short_output=False):
    space = '' if short_output else ' '
    newline = '' if short_output else '\n'

    if isinstance(grid, dict):
        # This is a (partial) solution
        for r in rows:
            for c in cols:
                v = grid[r+c]
                if len(v) == 1:
                    sys.stdout.write(v)
                else:
                    sys.stdout.write('.')
                if c != '9':
                    sys.stdout.write(space)
            sys.stdout.write(newline)
    else:
        indices = [(i,j) for i in range(9) for j in range(9)]
        for n, (i, j) in zip(grid, indices):
            sys.stdout.write(n if n in digits else '.')
            if j != 8:
                sys.stdout.write(space)
            else:
                sys.stdout.write(newline)

    sys.stdout.flush()

def grid_from_file(fp):
    grid = []
    while True:
        c = fp.read(1)
        if not c:
            raise EOFError

        if c.isspace():
            continue

        if c in digits:
            grid.append(c)
        else:
            grid.append('.')

        if len(grid) == 81:
            return grid

def main():
    parser = argparse.ArgumentParser(description='Sudoku solver (Python)')
    parser.add_argument('sudoku_files', type=str, nargs='*',
                        help='input file(s)')
    parser.add_argument('--short-output', '-s',
                        help='Use a shorter output format',
                        action='store_true')
    parser.add_argument('--timeit', metavar='ITERATIONS', type=int,
                        default=0)

    args = parser.parse_args()

    filenames = args.sudoku_files or ['-']

    for fn in filenames:
        if fn == '-':
            process_sudoku_file(sys.stdin, args.short_output, args.timeit)
        else:
            with open(fn, 'r') as fp:
                process_sudoku_file(fp, args.short_output, args.timeit)

    return 0

def process_sudoku_file(fp, short_output=False, timeit_iters=0):
    while True:
        try:
            grid = grid_from_file(fp)
            if not short_output:
                print('Sudoku:')
                print_grid(grid)
                print()

            if timeit_iters == 0:
                solved_grid = solve(grid)

                print_grid(solved_grid, short_output=short_output)
                print()
            else:
                t1 = time.perf_counter()
                for _ in range(timeit_iters):
                    solved_grid = solve(grid)
                t2 = time.perf_counter()
                dt_ms = (t2 - t1) * 1000

                print_grid(solved_grid, short_output=short_output)
                if short_output:
                    print(' %f' % (dt_ms/timeit_iters))
                else:
                    print('Running time %.2f s (%.2f ms per iteration)' %
                          (dt_ms, dt_ms/timeit_iters))

        except (ValueError, EOFError, IOError):
            return





if __name__ == '__main__':
    sys.exit(main())
